### PR TITLE
Added code to prevent NullReferebceExceptions if Area is not set.

### PR DESCRIPTION
**Pull request by ydenega**
Without this code Kendo UI grid that uses Razor view cannot function. For some reason AJAX calls to the controller are coming without Area properly set, and we are getting NullReferenceException in the Orchard code before we even reach the Controller.

### CODE REVIEW

File: src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs [hunk 1]
```diff
@@ -90,6 +90,11 @@ public class ExtensionManager : IExtensionManager
 
         public IExtensionInfo GetExtension(string extensionId)
         {
+            if (string.IsNullOrEmpty(extensionId))
```
Comment (sebastienros) [inline src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs]: Can you provide an explanation on how you reach this state?
Depending on your answer we might prefer to throw an exception instead.

File: src/OrchardCore/Orchard.DisplayManagement/LocationExpander/ThemeAwareViewLocationExpanderProvider.cs [hunk 2]
```diff
@@ -43,6 +43,11 @@ public void PopulateValues(ViewLocationExpanderContext context)
         public virtual IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context,
                                                                IEnumerable<string> viewLocations)
         {
+            if (string.IsNullOrEmpty(context.AreaName))
```
Comment (sebastienros) [inline src/OrchardCore/Orchard.DisplayManagement/LocationExpander/ThemeAwareViewLocationExpanderProvider.cs]: I assume this is empty for the same reason there is no extension id ...

**Comment by Jetski5822** [comment]
Is there no way to include `area` in the ajax calls from Kendo?

**Comment by sebastienros** [comment]
@Jetski5822 `area` is not something that is set from a client. When the url matches a route it's injected in the route data.

**Comment by sebastienros** [comment]
And by the way we might want to change this to use `module` instead, to be clearer.

**Comment by ydenega** [comment]
I have not found a way to force Area on Ajax calls grid makes to the Controller. I will provide you with an example shortly.

**Comment by ydenega** [comment]
OK, this is what I did to test it using Orchard code base:
1. Orchard.Cms.Web and Orchard.Users - I referenced 'Telerik.UI.for.AspNet.Core' (or Telerik.UI.for.AspNet.Core'.Trial) NuGet package.

2. In Orchard.Cms.Web\Startup.cs, ConfigureServices, I added
services.AddKendo();

3. In OrchardCore.Themes\TheAdmin\Views\Layout.cshtml, added kendo scripts and css:
    
    <link href="https://kendo.cdn.telerik.com/2017.2.504/styles/kendo.common-material.min.css" rel="stylesheet" type="text/css" />
    <link href="https://kendo.cdn.telerik.com/2017.2.504/styles/kendo.mobile.all.min.css" rel="stylesheet" type="text/css" />
    <link href="https://kendo.cdn.telerik.com/2017.2.504/styles/kendo.dataviz.min.css" rel="stylesheet" type="text/css" />
    <link href="https://kendo.cdn.telerik.com/2017.2.504/styles/kendo.material.min.css" rel="stylesheet" type="text/css" />
    <link href="https://kendo.cdn.telerik.com/2017.2.504/styles/kendo.dataviz.material.min.css" rel="stylesheet" type="text/css" />
    
    <script asp-name="jQuery" use-cdn="true" at="Head"></script>
    <script asp-name="kendo-jszip" use-cdn="true" at="Head"></script>
    <script asp-name="kendo-all-min" use-cdn="true" at="Head"></script>
    <script asp-name="kendo_aspnetmvc_min" use-cdn="true" at="Head"></script>

4. In OrchardCore.Modules\Orchard.Users\Views\Admin\Index.cshtml, I added a dummy grid:
		@(Html.Kendo().Grid<TelerikAspNetCoreApp2.Models.OrderViewModel>()
                .Name("grid")
                .Columns(columns =>
                {
                    columns.Bound(p => p.OrderID).Filterable(false);
                    columns.Bound(p => p.Freight);
                    columns.Bound(p => p.OrderDate).Format("{0:MM/dd/yyyy}");
                    columns.Bound(p => p.ShipName);
                    columns.Bound(p => p.ShipCity);
                    columns.Command(command => { command.Edit(); command.Destroy(); }).Width(160);
                })
		      .ToolBar(toolbar => toolbar.Create())
		      .Editable(editable => editable.Mode(GridEditMode.PopUp))
                .Pageable()
                .Sortable()
                .Scrollable()
                .Filterable()
                .HtmlAttributes(new { style = "height:550px;" })
                .DataSource(dataSource => dataSource
                    .Ajax()
                    .PageSize(20)
                    .Read(read => read.Action("Orders_Read", "Admin"))
                    .Update(read => read.Action("Orders_Read", "Admin"))
                    .Create(read => read.Action("Orders_Read", "Admin"))
                    .Destroy(read => read.Action("Orders_Read", "Admin"))
                    .ServerOperation(true)
                )
		)

5.  In Orchard.Users\Controllers\AdminController.cs, added Orders_Read method:
        public ActionResult Orders_Read([DataSourceRequest]DataSourceRequest request)
        {
            var result = Enumerable.Range(0, 50).Select(i => new OrderViewModel
            {
                OrderID = i,
                Freight = i * 10,
                OrderDate = new DateTime(2016, 9, 15).AddDays(i % 7),
                ShipName = "ShipName " + i,
                ShipCity = "ShipCity " + i
            });

            var dsResult = result.ToDataSourceResult(request);
            return Json(dsResult);
        }

6. Added OrderViewModel.cs:
using System;
using System.ComponentModel.DataAnnotations;
using Newtonsoft.Json;

namespace Orchard.Users.ViewModels
{
    [JsonObject(MemberSerialization.OptIn)]
    public class OrderViewModel
    {
        [JsonProperty("OrderID")]
        public int OrderID
        {
            get;
            set;
        }

        [JsonProperty("Freight")]
        public decimal? Freight
        {
            get;
            set;
        }

        [Required]
        [JsonProperty("OrderDate")]
        public DateTime? OrderDate
        {
            get;
            set;
        }

        [JsonProperty("ShipCity")]
        public string ShipCity
        {
            get;
            set;
        }

        [JsonProperty("ShipName")]
        public string ShipName
        {
            get;
            set;
        }
    }
}

File: src/OrchardCore/Orchard.DisplayManagement/LocationExpander/ThemeAwareViewLocationExpanderProvider.cs (hunk 2 above)
Comment (ydenega) [inline src/OrchardCore/Orchard.DisplayManagement/LocationExpander/ThemeAwareViewLocationExpanderProvider.cs]: I guess so. They always go in pairs.

**Comment by ydenega** [comment]
I put a grid directly into existing Orchard code to make sure this is really an Orchard issue. That's why I have not created a separate module for testing. This is obviously just a test case. In our code we use custom modules, and it also fails.

File: src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs (hunk 1 above)
Comment (ydenega) [inline src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs]: If you throw an exception, it probably will not help. As a matter of fact the exception is thrown down the stack when a null area object is referenced, and this exception terminates pipeline processing. As a result the controller is never reached.

Comment (sebastienros) [inline src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs]: I mean to throw an exception before this is an unexpected and unsupported state.

Comment (sebastienros) [inline src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs]: But please, can you provide an explanation on how you reach this state?

**Comment by ydenega** [comment]
Oh, one more thing:
You need to send AntForgeryToken back, so it needs to be like this:

.Read(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))
.Update(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))
.Create(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))
.Destroy(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))

although in our case it probably does not matter. If I remember correctly, the failure happens NOT on the call to AJAX method, but rather on rendering Razor view in the first place.  If you have CRUD actions defined in the view, I believe Orchard watches the response stream, and tries to do something with it. That's where it invokes the code in question that fails.

**Comment by ydenega** [comment]
Hello, please let me know if info I provided is sufficient. Is there anything else I can help you with regarding this issue?

File: src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs (hunk 1 above)
Comment (ydenega) [inline src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs]: OK, this is what I did to test it using Orchard code base:

Orchard.Cms.Web and Orchard.Users - I referenced 'Telerik.UI.for.AspNet.Core' (or Telerik.UI.for.AspNet.Core'.Trial) NuGet package.

In Orchard.Cms.Web\Startup.cs, ConfigureServices, I added
services.AddKendo();

In OrchardCore.Themes\TheAdmin\Views\Layout.cshtml, added kendo scripts and css:

<script asp-name="jQuery" use-cdn="true" at="Head"></script> <script asp-name="kendo-jszip" use-cdn="true" at="Head"></script> <script asp-name="kendo-all-min" use-cdn="true" at="Head"></script> <script asp-name="kendo_aspnetmvc_min" use-cdn="true" at="Head"></script>
In OrchardCore.Modules\Orchard.Users\Views\Admin\Index.cshtml, I added a dummy grid:
@(Html.Kendo().Grid<TelerikAspNetCoreApp2.Models.OrderViewModel>()
.Name("grid")
.Columns(columns =>
{
columns.Bound(p => p.OrderID).Filterable(false);
columns.Bound(p => p.Freight);
columns.Bound(p => p.OrderDate).Format("{0:MM/dd/yyyy}");
columns.Bound(p => p.ShipName);
columns.Bound(p => p.ShipCity);
columns.Command(command => { command.Edit(); command.Destroy(); }).Width(160);
})
.ToolBar(toolbar => toolbar.Create())
.Editable(editable => editable.Mode(GridEditMode.PopUp))
.Pageable()
.Sortable()
.Scrollable()
.Filterable()
.HtmlAttributes(new { style = "height:550px;" })
.DataSource(dataSource => dataSource
.Ajax()
.PageSize(20)
.Read(read => read.Action("Orders_Read", "Admin"))
.Update(read => read.Action("Orders_Read", "Admin"))
.Create(read => read.Action("Orders_Read", "Admin"))
.Destroy(read => read.Action("Orders_Read", "Admin"))
.ServerOperation(true)
)
)

In Orchard.Users\Controllers\AdminController.cs, added Orders_Read method:
public ActionResult Orders_Read([DataSourceRequest]DataSourceRequest request)
{
var result = Enumerable.Range(0, 50).Select(i => new OrderViewModel
{
OrderID = i,
Freight = i * 10,
OrderDate = new DateTime(2016, 9, 15).AddDays(i % 7),
ShipName = "ShipName " + i,
ShipCity = "ShipCity " + i
});

    var dsResult = result.ToDataSourceResult(request);
    return Json(dsResult);
}
Added OrderViewModel.cs:
using System;
using System.ComponentModel.DataAnnotations;
using Newtonsoft.Json;

namespace Orchard.Users.ViewModels
{
[JsonObject(MemberSerialization.OptIn)]
public class OrderViewModel
{
[JsonProperty("OrderID")]
public int OrderID
{
get;
set;
}

    [JsonProperty("Freight")]
    public decimal? Freight
    {
        get;
        set;
    }

    [Required]
    [JsonProperty("OrderDate")]
    public DateTime? OrderDate
    {
        get;
        set;
    }

    [JsonProperty("ShipCity")]
    public string ShipCity
    {
        get;
        set;
    }

    [JsonProperty("ShipName")]
    public string ShipName
    {
        get;
        set;
    }
}
}

I put a grid directly into existing Orchard module to make sure this is really an Orchard issue. This is obviously just a test case. In our code we use custom modules, and it also fails.

You also need to send AntForgeryToken back, so it needs to be like this:

.Read(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))
.Update(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))
.Create(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))
.Destroy(read => read.Action("Orders_Read", "Admin").Data("kendo.antiForgeryTokens"))

although in our case it probably does not matter. If I remember correctly, the failure happens NOT on the call to AJAX method, but rather on rendering Razor view in the first place. If you have CRUD actions defined in the view, I believe Orchard watches the response stream, and tries to do some procession based on the content if the Response. That's where it invokes the code in question that fails.

**Comment by sfmskywalker** [comment]
IT looks like something isn't right with your PR - way too many commits from others. I recommend creating a new PR with only _your_ changes.

**Comment by sebastienros** [comment]
Fixed in https://github.com/OrchardCMS/OrchardCore/commit/e54c9315b67691fc891d1fd6edd5724cb0eff2d8

**Pull request closed without merging**